python_modules/libraries/dagster-shared/dagster_shared/serdes/serdes.py

```diff
@@ -1101,6 +1127,169 @@ def _wrap_object(
 U_PackableValue = TypeVar("U_PackableValue", bound=PackableValue, default=PackableValue)
 
 
+def lazy_deserialized(cls):
+    class MetaWithLazyPackedDictInstanceCheck(type(cls)):
+        def __instancecheck__(self, instance):
+            if isinstance(instance, LazyPackedDict):
+                deserializer = instance.get_whitelist_map().object_deserializers.get(
+                    instance.get_klass_name()
+                )
+                if deserializer:
+                    return deserializer.klass.__name__ == cls.__name__
+                return False
+            return super().__instancecheck__(instance)
+
+    # Construct a *new class* with the new metaclass
+    return MetaWithLazyPackedDictInstanceCheck(
+        cls.__name__, (cls, *cls.__bases__), dict(cls.__dict__)
+    )
+
+
+counter = {"count": 0}
+
+
+class _UnpackedSentinel:
+    """Sentinel object to mark items that haven't been unpacked yet."""
+
+    _cached_value: Optional[UnpackedValue] = None
+
+    def __init__(
+        self,
+        json_value: JsonSerializableValue,
+        whitelist_map: WhitelistMap,
+        context: UnpackContext,
+    ):
+        counter["count"] += 1
+        self.id = counter["count"]
+        self.json_value = json_value
+        self._whitelist_map = whitelist_map
+        self._context = context
+
+    def get_value(self) -> UnpackedValue:
+        # Check if we already have the cached unpacked value
+        if self._cached_value is not None:
+            return self._cached_value
+
+        # Unpack and cache the value
+        unpacked_value = _unpack_value(self.json_value, self._whitelist_map, self._context)
+        self._cached_value = unpacked_value
+
+        # TODO Delete json_value safely
```

Comment:
For some reason when I deleted it here `del self.json_value` I would somehow end up on line 1174 again. Super weird.